### ISSUE TITLE
Out of range arguments for blockchain.block.headers

### ISSUE THREAD

**Issue by dagurval** [dagurval/electrum-cash-protocol#8]
What should happen if:
- `count` is 0
- `count + start_height` is more than height of current tip
- `start_height` is above current tip
- `count` is 0 and `cp_height` is defined

Also:
- Is `start_height == 0` genesis?

**Comment by cculianu** [comment]
Hmm good point. I'll look at what Fulcrum and EX does and/or what Electron Cash expects and we can go from there as far as what we can define that won't break things, I guess...

**Comment by dagurval** [comment]
Agreed, best is to document existing behavior

**Comment by cculianu** [comment]
yeah -- I agree. If this were a new protocol from scratch we could do the "ideal" thing but since we have code out there in the wild in multiple libs and clients -- best to not redefine.  I'll get you an answer on this right now, since I have the code open.

**Comment by cculianu** [comment]
Fulcrum and ElectrumX (I made sure they behave the same to save myself headaches):

- count is 0
  - Returns `0` results with an empty hex string. Example:
    `{"id":1,"jsonrpc":"2.0","result":{"count":0,"hex":"","max":2016}}`

- count + start_height > latest_height.
  - Depends entirely on the `cp_height` parameter being present (nonzero), or not being present (and/or zero).
  - In the simple case, with no `cp_height`: Returns what it can up until tip. Does not return error. May even return `0` results (as if count=0 above) if `start_height` is > tip_height.
  - In the `cp_height` case it *DOES* return an error if the count is out of range, or if start_height is out of range, or if cp_height is out of range. Both EX and Fulcrum behave the same way here, and their error message is very explicit..  E.g.:
    - `{"error":{"code":1,"message":"header height + (count - 1) 1360263 must be <= cp_height 1360250 which must be <= chain height 1360263"},"id":1,"jsonrpc":"2.0"}`
- start_height is above current tip:
  - Behavior depends entirely on `cp_height` being present and nonzero or not.  
      - In the simple case (cp_height=0 or not present): Returns `0` results (see first bullet point above).  
      - In the `cp_height` case, returns the same error message as the previous bullet point above, about being out of range.
- `count` is `0` and `cp_height` is defined:
  - Returns the same thing as simple `count` is `0` e.g.:
    - `{"id":1,"jsonrpc":"2.0","result":{"count":0,"hex":"","max":2016}}`
- If `start_height == 0` (genesis):
   - Returns the genesis header. No special case here.

**Comment by dagurval** [comment]
Wow, thanks for digging that out. I'll try to write it to the spec in a way that doesn't scare the reader :D

**Comment by cculianu** [comment]
Ha ha ha yeah. I appreciate your taking the time to do this diligently. :)

**Comment by cculianu** [comment]
Do let me know if this doesn't jive 100% with what Electrs does and we can take it from there..
